articles/string-manipulation/index.md

```diff
@@ -0,0 +1,125 @@
+---
+layout: engineering-education
+status: publish
+published: true
+url: /engineering-education/string-manipulation/
+title: String Manipulation
+description: The article discusses how to modify strings in the C programming language. This includes discussing functions from the string library.
+author: nimra-aftab
+date: 2020-09-11T00:00:00-07:00
+topics: [Languages]
+excerpt_separator: <!--more-->
+images:
+
+  - url: /engineering-education/string-manipulation/hero.jpg
+    alt: string manipulation image example
+---
+Recall that there are two types of strings in the C programming language. String *variables* can be modified, but string *literals* cannot. This article will discuss how to safely manipulate string variables.
+<!--more-->
+
+### String Basics
+String *variables* are an array of characters in 'C'. Strings are always null terminated, so the array needs space for the string **and** the null character (`\0`).
+
+```c
+/* There are two ways to declare a string variable */
+char mystr[4] = "bye"; // '\0' is appended automatically to the end
+char mystr2[4] = {'b', 'y', 'e', '\0'};
+```
+We can change characters in a string by modifying the array elements:
+``` c
+mystr[0] = 'e'; // mystr is now "eye" instead of "bye"
+```
+
+### Identifying a Common Mistake
+Suppose we wanted to copy the contents of a string. Here is what beginners often write:
+
+```c
+char mystr[6] = "hello";
+char newstr[6];
+newstr = mystr; // THIS IS INCORRECT.
+```
+Notice that `newstr ` is an array name. It is **illegal** to use an array name on the left side of the assignment operator (`=`). The only exception to this rule is when we initialize a string **at the same time** as we declare it. So `char mystr[6] = "hello"` is acceptable, but `newstr = mystr` is not.
+
+Luckily, there is a string function in 'C' that can help us copy strings.
+
+### Introducing the String Library
+The string library has functions that can help us manipulate strings. You'll need to add `#include <string.h>` at the beginning of your program to access these functions.
+
+Let's continue to solve the problem of copying the contents of a string. There are two functions that may be able to help us: `strcpy` and `strncpy`. We will look at `strcpy` first.
+
+```c
+#include <string.h> // for strcpy()
+#include <stdio.h> // for printf()
+int main(){
+    char mystr[6] = "hello";    
+    char newstr[6];
+    /* copy the contents of 'mystr' into 'newstr',
+    including the null char */
+    strcpy(newstr, mystr);
+    printf("Now 'newstr' contains %s \n", newstr);
+}
+```
```

Comment:
Best practice is to write this code up in repl.it. From there we can imbed the code runner in the article. You're ready for final review, but if you can do this for the next article, it would be great!